containers/Dockerfile.buildenv

```diff
@@ -309,3 +280,103 @@ RUN wget https://downloadmirror.intel.com/684899/sde-external-9.0.0-2021-11-07-l
 
 # Remove the apt-get cache in order to reduce image size
 RUN apt-get -y clean
+
+FROM dev as ci
+
+# Install all the compilers we are going to test on CI
+RUN apt-get update -y \
+    && apt-get install -y gcc-10 g++-10 gfortran-10 \
+                          gcc-11 g++-11 gfortran-11
+
+# Install clang-11 and clang-13
+RUN wget -O - https://apt.llvm.org/llvm-snapshot.gpg.key | apt-key add - \
+    && add-apt-repository 'deb http://apt.llvm.org/focal/ llvm-toolchain-focal-11 main' \
+    && apt-get update -y && apt-get install -y clang-11 \
+    && wget -O - https://apt.llvm.org/llvm-snapshot.gpg.key | apt-key add - \
+    && add-apt-repository 'deb http://apt.llvm.org/focal/ llvm-toolchain-focal-13 main' \
+    && apt-get update -y && apt-get install -y clang-13
+
+# `futures` is needed for `yapf` parallelization on Python 2.
+RUN apt-get update -y \
+    && apt-get install -y curl python2 python2-dev \
+    && curl https://bootstrap.pypa.io/pip/2.7/get-pip.py --output get-pip.py \
+    && python2 get-pip.py && rm get-pip.py \
+    && pip2 --no-cache-dir install pybind11~=2.6.1 numpy scipy h5py matplotlib \
+      pyyaml \
+    && pip2 --no-cache-dir install futures unittest2
+
+# Also install our minimum required CMake version so we can test it on CI.
+RUN wget https://github.com/Kitware/CMake/releases/download/v3.12.4/cmake-3.12.4.tar.gz \
+    && tar -xzf cmake-3.12.4.tar.gz \
+    && cd cmake-3.12.4 \
+    && mkdir -p /opt/local/cmake/3.12.4 \
+    && ./bootstrap --prefix=/opt/local/cmake/3.12.4 \
+    && make $PARALLEL_MAKE_ARG\
+    && make install \
+    && cd .. && rm -rf cmake*
+
+# We build both Clang and GCC versions of Charm++ so that all our tests can
+# use the same build environment.
+WORKDIR /work/charm_7_0_0
+RUN ./build LIBS multicore-linux-x86_64 clang \
+      ${PARALLEL_MAKE_ARG} -g -O2 --build-shared\
+    && rm -rf /work/charm_7_0_0/doc /work/charm_7_0_0/examples
+
+# Remove the apt-get cache in order to reduce image size
+RUN apt-get -y clean
+
+WORKDIR /work
+
+
+# We don't inherit from ci because that has lots of unnecessary compilers that
+# we don't need and would only increase the size of the image. We inheret from
+# the remote image rather than the local dev because it is faster on release
+# CI to just pull the remote dev image and buiild demo rather than having to
+# build all of dev and demo combined.
```

Comment:
There's a way to fetch the image and then make sure that Docker uses it as cache when building. Maybe add a note here to try this?

Reply:
Is that not what this does? Pulls the remote image and uses that as cache and builds `demo` on top of it?

If it isn't, could you refer me to what you had in mind?

Reply:
You would normally base `demo FROM dev`, not from the remote image that must be pushed before. My understanding is you based on the remote image so the post-release workflow doesn't have to rebuild `dev`. And my comment is: there's a way to pre-fetch `dev` so Docker doesn't need to rebuild it. My suggestion is to add a comment so it's clear this is something we could try, but didn't bother just yet.

Reply:
Ah ok I understand. I added in a couple lines to the fixup explaining this.